### PR TITLE
Documentation: grpc-gateway

### PR DESCRIPTION
**Pull request by gyuho**
- Addressing https://github.com/coreos/etcd/issues/5686
 - generate swagger API definitions in `genproto.sh`
 - document how to use `grpc-gateway`

### CODE REVIEW

**Comment by gyuho** [comment]
/cc @xiang90 @heyitsanthony 

Please review.

This is first draft. I will add more contents based on the feedback.

Thanks!

File: Documentation/demo.md [hunk 1]
```diff
@@ -284,3 +284,30 @@ etcdctl --write-out=table --endpoints=$ENDPOINTS snapshot status my.db
 | c55e8b8 |        9 |         13 | 25 kB      |
 +---------+----------+------------+------------+
 ```
+
+
+## Migrate
```
Comment (xiang90) [inline Documentation/demo.md]: make this a separate pr?

File: Documentation/dev-guide/api_grpc_gateway.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
+
+## Why grpc-gateway
+
+etcd v3 uses [gRPC][grpc]. Most recommended is to use etcd [Go client][go-client] and [etcdctl][etcdctl]. However, some languages are not supported by gRPC. This is where [grpc-gateway][grpc-gateway] steps in. It builds a RESTful proxy based on the gRPC services, which can serve HTTP, JSON requests. 
+
+[go-client]: https://github.com/coreos/etcd/tree/master/clientv3
```
Comment (xiang90) [inline Documentation/dev-guide/api_grpc_gateway.md]: move these links to the bottom of the doc

File: Documentation/dev-guide/api_grpc_gateway.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+ETCDCTL_API=3 etcdctl --endpoints=localhost:2379 get foo
+```
+
+
+## Swagger
+
+Documentation...
+
+https://github.com/coreos/etcd/blob/master/etcdserver/etcdserverpb/rpc.swagger.json
```
Comment (xiang90) [inline Documentation/dev-guide/api_grpc_gateway.md]: use relative path

File: Documentation/dev-guide/api_grpc_gateway.md (hunk 2 above)
Comment (xiang90) [inline Documentation/dev-guide/api_grpc_gateway.md]: use relative path if the link is inside etcd repo

File: Documentation/dev-guide/api_grpc_gateway.md (hunk 3 above)
Comment (gyuho) [inline Documentation/dev-guide/api_grpc_gateway.md]: Ok will fix. Thanks.

File: Documentation/dev-guide/api_grpc_gateway.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
+
+## Why grpc-gateway
+
+etcd v3 uses [gRPC][grpc]. Most recommended is to use etcd [Go client][go-client] and [etcdctl][etcdctl]. However, some languages are not supported by gRPC. This is where [grpc-gateway][grpc-gateway] steps in. It builds a RESTful proxy based on the gRPC services, which can serve HTTP and JSON requests.
+
+
+## How grpc-gateway
+
+gRPC messages are in Protocol Buffer, serialized binary data. So requests to grpc-gateway need to be base64-encoded.
```
Comment (xiang90) [inline Documentation/dev-guide/api_grpc_gateway.md]: The gateway accepts the [JSON mapping](https://developers.google.com/protocol-buffers/docs/proto3#json) for the messages defined in [Protocol Buffer](link to our protobuf api doc) format.

Comment (xiang90) [inline Documentation/dev-guide/api_grpc_gateway.md]: Note that key, value are defined as bytes in Protocol Buffer, thus need to be encoded as base64 string.

File: Documentation/dev-guide/api_grpc_gateway.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+## Swagger
+
+Generated [Swapper][swagger] API definitions can be found at [rpc.swagger.json][swagger-doc].
+
+[go-client]: https://github.com/coreos/etcd/tree/master/clientv3
+[etcdctl]: https://github.com/coreos/etcd/tree/master/etcdctl
+[grpc]: http://www.grpc.io/
+[grpc-gateway]: https://github.com/gengo/grpc-gateway
+[swagger]: http://swagger.io/
+[swagger-doc]: ../../etcdserver/etcdserverpb/rpc.swagger.json
```
Comment (xiang90) [inline Documentation/dev-guide/api_grpc_gateway.md]: i think swagger can generate a good browseable page?

**Comment by gyuho** [comment]
@xiang90 Just fixed. And did you mean this http://swagger.io/swagger-ui? Not sure how we want to integrate this?

**Comment by xiang90** [comment]
@gyuho something like http://kubernetes.io/kubernetes/third_party/swagger-ui/#!/api%2Fv1/listEndpoints

File: etcdserver/etcdserverpb/rpc.swagger.json [hunk 6]
```diff
@@ -0,0 +1,1239 @@
+{
```
Comment (xiang90) [inline etcdserver/etcdserverpb/rpc.swagger.json]: move this to dev-guide/apispec/swagger/

**Comment by gyuho** [comment]
@xiang90 Just fixed. The generated path got more verbose because swagger-gen includes the parent directory. PTAL.

And for the Kubernetes doc, we need to set up separate server to serve those docs for etcd, I think. Once we stabilize the API, I think we can set up doc website or something?

**Comment by xiang90** [comment]
@gyuho We can set it up next week on api.etcd.io/v3/swagger

**Comment by xiang90** [comment]
lgtm. defer to @heyitsanthony @joshix for doc checking.

**Comment by gyuho** [comment]
Sure, sgtm.

File: Documentation/dev-guide/apispec/swagger/etcdserver/etcdserverpb/rpc.swagger.json [hunk 7]
```diff
@@ -0,0 +1,1239 @@
+{
```
Comment (xiang90) [inline Documentation/dev-guide/apispec/swagger/etcdserver/etcdserverpb/rpc.swagger.json]: this is still not good location. can we just put it under /apispec/swagger instead of apispec/swagger/etcdserver/etcdserverpb/

Comment (gyuho) [inline Documentation/dev-guide/apispec/swagger/etcdserver/etcdserverpb/rpc.swagger.json]: yeah... let me play with the gen more.

**Comment by gyuho** [comment]
Just changed the script to move swagger doc manually. It's a bit ugly but we run it file by file anyway, so it's not that bad.

File: Documentation/dev-guide/api_grpc_gateway.md [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+bar is 'YmFy'
+COMMENT
+
+curl -L http://localhost:2379/v3alpha/kv/put \
+	-X POST -d '{"key": "Zm9v", "value": "YmFy"}'
+
+curl -L http://localhost:2379/v3alpha/kv/range \
+	-X POST -d '{"key": "Zm9v"}'
+
+ETCDCTL_API=3 etcdctl --endpoints=localhost:2379 get foo
```
Comment (xiang90) [inline Documentation/dev-guide/api_grpc_gateway.md]: why the etcdctl command is here?

Comment (gyuho) [inline Documentation/dev-guide/api_grpc_gateway.md]: I put it there to show that it was successfully written. But now it looks un-related. Will remove.

**Comment by gyuho** [comment]
@xiang90 just removed the unnecessary etcdctl command.

File: Documentation/dev-guide/api_grpc_gateway.md [hunk 9]
```diff
@@ -0,0 +1,38 @@
+
+## Why grpc-gateway
+
+etcd v3 uses [gRPC][grpc]. Most recommended is to use etcd [Go client][go-client] and [etcdctl][etcdctl]. However, some languages are not supported by gRPC. This is where [grpc-gateway][grpc-gateway] steps in. It builds a RESTful proxy based on the gRPC services, which can serve HTTP and JSON requests.
```
Comment (heyitsanthony) [inline Documentation/dev-guide/api_grpc_gateway.md]: etcd v3 uses [gRPC][grpc] for its messaging protocol. The etcd project includes a gRPC-based [Go client][go-client] and a command line utility, [etcdctl][etcdctl], for communicating with an cluster etcd through gRPC. For languages with no gRPC support, etcd provides a JSON [grpc-gateway][grpc-gateway]. This gateway serves a RESTful proxy that translates HTTP/JSON requests into gRPC messages.

File: Documentation/dev-guide/api_grpc_gateway.md [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+
+## How grpc-gateway
```
Comment (heyitsanthony) [inline Documentation/dev-guide/api_grpc_gateway.md]: s/How/Using

**Comment by gyuho** [comment]
@heyitsanthony Just fixed. PTAL.

Thanks!

File: Documentation/dev-guide/api_grpc_gateway.md [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+The gateway accepts the [JSON mapping][json-mapping] for the messages defined in [Protocol Buffer][api-ref] format. Note that key, value are defined as bytes in Protocol Buffer, thus need to be encoded as base64 string.
```
Comment (heyitsanthony) [inline Documentation/dev-guide/api_grpc_gateway.md]: The gateway accepts a [JSON mapping][json-mapping] for etcd's [protocol buffer][api-ref] message definitions. Note that `key` and `value` fields are defined as byte arrays and therefore must be base64 encoded in JSON.

File: Documentation/dev-guide/api_grpc_gateway.md [hunk 12]
```diff
@@ -0,0 +1,38 @@
+
+## Why grpc-gateway
+
+etcd v3 uses [gRPC][grpc] for its messaging protocol. The etcd project includes a gRPC-based [Go client][go-client] and a command line utility, [etcdctl][etcdctl], for communicating with an cluster etcd through gRPC. For languages with no gRPC support, etcd provides a JSON [grpc-gateway][grpc-gateway]. This gateway serves a RESTful proxy that translates HTTP/JSON requests into gRPC messages.
```
Comment (heyitsanthony) [inline Documentation/dev-guide/api_grpc_gateway.md]: s/cluster etcd/etcd cluster/

**Comment by heyitsanthony** [comment]
lgtm

**Comment by gyuho** [comment]
all fixed. merging in greens. thanks.
